scripts/question_answering/question_answering.py

```diff
@@ -0,0 +1,324 @@
+
+        with self.name_scope():
+            self._char_dense_embedding = nn.Embedding(input_dim=len(char_vocab),
+                                                      output_dim=8,
+                                                      weight_initializer=initializer.Uniform(0.001))
+            self._dropout = nn.Dropout(rate=dropout)
+            self._char_conv_embedding = ConvolutionalEncoder(
+                embed_size=8,
+                num_filters=(100,),
+                ngram_filter_sizes=(5,),
+                num_highway=None,
+                conv_layer_activation='relu',
+                output_size=None
+            )
+
+            self._word_embedding = nn.Embedding(prefix='predefined_embedding_layer',
+                                                input_dim=len(word_vocab),
+                                                output_dim=embedding_size)
+
+            self._highway_network = Highway(2 * embedding_size, num_layers=highway_nlayers)
+            self._contextual_embedding = LSTM(hidden_size=embedding_size,
+                                              num_layers=contextual_embedding_nlayers,
+                                              bidirectional=True, input_size=2 * embedding_size,
+                                              dropout=dropout)
+
+    def init_embeddings(self, grad_req='null'):
+        """Initialize words embeddings with provided embedding values
+
+        Parameters
+        ----------
+        grad_req: str
+            How to treat gradients of embedding layer
+        """
+        self._word_embedding.weight.set_data(self._word_vocab.embedding.idx_to_vec)
+        self._word_embedding.collect_params().setattr('grad_req', grad_req)
+
+    def hybrid_forward(self, F, w, c, *args):
+        # pylint: disable=arguments-differ,unused-argument,missing-docstring
+        word_embedded = self._word_embedding(w)
+        char_level_data = self._char_dense_embedding(c)
+        char_level_data = self._dropout(char_level_data)
+
+        # Transpose to put seq_len first axis to iterate over it
+        char_level_data = F.transpose(char_level_data, axes=(1, 2, 0, 3))
+
+        def convolute(token_of_all_batches, _):
+            return self._char_conv_embedding(token_of_all_batches), []
+        char_embedded, _ = F.contrib.foreach(convolute, char_level_data, [])
+
+        # Transpose to TNC, to join with character embedding
+        word_embedded = F.transpose(word_embedded, axes=(1, 0, 2))
+        highway_input = F.concat(char_embedded, word_embedded, dim=2)
+
+        def highway(token_of_all_batches, _):
+            return self._highway_network(token_of_all_batches), []
+
+        # Pass through highway, shape remains unchanged
+        highway_output, _ = F.contrib.foreach(highway, highway_input, [])
+
+        # Transpose to TNC - default for LSTM
+        ce_output = self._contextual_embedding(highway_output)
+        return ce_output
+
+
+class BiDAFModelingLayer(HybridBlock):
```

Comment:
This class doesn't seem to add much value.

Reply:
Yes, but I would prefer to leave it as there is such an entity in the paper: picture and a reference block of text. So, for the ease of understanding, I wouldn't remove it.

Reply:
You can do that in the BiDAFModel class since `self.modeling_layer = gluon.rnn.LSTM(...)` is as readable as `self.modeling_layer = BiDAFModelingLayer()`

Reply:
Removed the class